### DISCUSSION TITLE
crash of ssr server

### DISCUSSION THREAD

**Discussion by s-light** [quasarframework/quasar#15563] [CLI - SSR mode] [answered]
dev server in ssr mode:
`quasar dev -m ssr --debug`
 
```
...
[Vue warn]: Unhandled error during execution of watcher callback 
  at <JobPage onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > >
[Vue warn]: Unhandled error during execution of setup function 
  at <JobPage onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > >
projectname/node_modules/ouch/handler/PrettyPageHandler.js:328
            SCRIPT_FILE: require.main.filename,
                                      ^

TypeError: Cannot read properties of undefined (reading 'filename')
    at PrettyPageHandler.__getServerAndRequestInfo (projectname/node_modules/ouch/handler/PrettyPageHandler.js:328:39)
    at PrettyPageHandler.handle (projectname/node_modules/ouch/handler/PrettyPageHandler.js:143:41)
    at next (projectname/node_modules/ouch/Ouch.js:137:25)
    at projectname/node_modules/ouch/Ouch.js:141:9
    at Ouch.handleException (projectname/node_modules/ouch/Ouch.js:142:7)
    at Object.renderError [as error] (projectname/kkjobs/node_modules/@quasar/app-vite/lib/modes/ssr/ssr-devserver.js:44:16)
    at projectname/.quasar/ssr/compiled-dev-webserver.js:80:19

Node.js v18.15.0
```
i have a bug in my app - but this should not crash the dev server or???

**Comment by rstoenescu** [comment]
@s-light  Can you offer us a reproduction repo, pls? There is no other way for us to be able to investigate the issue.

**Reply by s-light** [reply]
done: 
https://github.com/s-light/quasar_ssr_crash_test

-  `git clone https://github.com/s-light/quasar_ssr_crash_test.git`
- `yarn`
- `quasar dev -m ssr --debug`
- in webbrowser open a new tab to `http://localhost:9100/`
- app is shown.
- in webbrowser open a new tab to `http://localhost:9100/job/5`
- server crashed. (have a look in your terminal)
- i know there is a bug...  that was the original point. but should the dev-server crash on this?

hope that helps...

**Reply by rstoenescu** [reply]
Hi,

Investigated but unfortunately this is due to Vue architecture itself. There is nothing that we can do in Quasar about this particular issue, unless I am missing something major.

**Reply by s-light** [reply]
thanks for looking into it.
does it make sens to report this somewhere to Vue?
or is it more a -it is just as it is thing?
maybe give a hint in the documentation? or iam the only one wondering? ;-)

**Reply by freddy38510** [reply]
@rstoenescu I faced the same issue with the ssg extension I'm developping.

This is related to the Ouch package. The [pretty page handler](https://github.com/quorrajs/Ouch/blob/master/handler/PrettyPageHandler.js#L328) is trying to access an undefined property `require.main`.

This issue is only affected by the use of the new `@quasar/cli` v2 package, because its entry point is **no longer** a CommonJS module.

A quote from the [Nodejs documentation](https://nodejs.org/api/modules.html#accessing-the-main-module):
> When the entry point is not a CommonJS module, require.main is undefined, and the main module is out of reach.

To solve this I extended the pretty page handler like this:

```js
// extend-pretty-page-handler.js

const os = require('os');
const url = require('url');

module.exports = function extendPrettyPageHandler(PrettyPageHandler) {
  PrettyPageHandler.prototype.__getServerAndRequestInfo = function __getServerAndRequestInfo() {
    const req = this.__request;

    if (req === null) {
      return {};
    }

    return {
      REMOTE_ADDR: (req.headers['x-forwarded-for'] || '').split(',')[0] || req.connection.remoteAddress,
      REMOTE_PORT: req.connection.remotePort,
      SERVER_SOFTWARE: `NodeJS ${process.versions.node} ${os.type()}`,
      SERVER_PROTOCOL: `${this.__getRequestProtocol()}/${req.httpVersion}`,
      REQUEST_URI: req.url,
      REQUEST_METHOD: req.method,
      /**
       * ensure to correctly access the main module that may be missing
       * when using @quasar/cli v2.0.0 because its entry point is not a CommonJS module
       *
       * @see https://nodejs.org/api/modules.html#accessing-the-main-module
       *
       * fallback to env variable then @quasar/app-vite main script
       */
      SCRIPT_FILE: require.main?.filename || process.env._ || require.resolve('@quasar/app-vite/bin/quasar'),
      PATH_INFO: url.parse(req.url).pathname,
      QUERY_STRING: url.parse(req.url).query,
      HTTP_HOST: req.headers.host,
      HTTP_CONNECTION: req.headers.connection,
      HTTP_CACHE_CONTROL: req.headers['cache-control'],
      HTTP_ACCEPT: req.headers.accept,
      HTTP_USER_AGENT: req.headers['user-agent'],
      HTTP_DNT: req.headers.dnt,
      HTTP_ACCEPT_ENCODING: req.headers['accept-encoding'],
      HTTP_ACCEPT_LANGUAGE: req.headers['accept-language'],
      HTTP_COOKIE: req.headers.cookie,
    };
  };

  return PrettyPageHandler;
};
```

```js
const Ouch = require('ouch');
const extendPrettyPageHandler = require('./extend-pretty-page-handler');

Ouch.handlers.PrettyPageHandler = extendPrettyPageHandler(Ouch.handlers.PrettyPageHandler);

const ouchInstance = new Ouch().pushHandler(
  new Ouch.handlers.PrettyPageHandler('orange', null, 'sublime'),
);
```

**Reply by rstoenescu** [reply]
Not sure how I missed this, but you're right. Investigating.

**Reply by rstoenescu** [reply]
Ouch package seems unmaintained. Was looking to write our own error handling page package anyway, so I guess this is the perfect opportunity to finally do it. Will be available in the next q/app-vite & q/app-webpack.

**Reply by rstoenescu** [reply]
Will be a nice change to also make these entries work.

![image](https://user-images.githubusercontent.com/4665641/235598945-85fb16ff-1edc-4b68-9e87-565a369d6fb0.png)

**Comment by rstoenescu** [comment, marked as answer]
Just wrote a new package for Quasar which will handle the 500 errors on SSR for dev mode.
A sneak peak:

![image](https://user-images.githubusercontent.com/4665641/236443670-d7c1ec4e-6c0f-4846-98f8-61987baa7684.png)

**Reply by rstoenescu** [reply]
Will be available in the next q/app-vite and q/app-webpack.
And it will also solve the initial issue with CLI crashing.

**Reply by rstoenescu** [reply]
q/app-vite 1.3.0 and q/app-webpack 3.8.0 contain this fix/feat. Just released them.

**Reply by freddy38510** [reply]
Wow, it looks really beautiful! Impressive work, well done @rstoenescu !
